### PR TITLE
Reduce the size of static data in std_unicode::tables

### PR DESCRIPTION
**Pull request by SimonSapin**
`BoolTrie` works well for sets of code points spread out through most of Unicode’s range, but is uses a lot of space for sets with few, mostly low, code points.

This switches a few of its instances to a similar but simpler trie data structure.

CC @raphlinus, who wrote the original `BoolTrie`.

## Before

`size_of::<BoolTrie>()` is 1552, which is added to `table.r3.len() * 8 + t.r5.len() + t.r6.len() * 8`:

* `Cc_table`: 1632
* `White_Space_table`: 1656
* `Pattern_White_Space_table`: 1640
* Total: 4928 bytes

## After

`size_of::<SmallBoolTrie>()` is 32, which is added to `t.r1.len() + t.r2.len() * 8`:

* `Cc_table`: 51
* `White_Space_table`: 273
* `Pattern_White_Space_table`: 193
* Total: 517 bytes

## Difference

Every Rust program with `std` statically linked should be about 4 KB smaller.

### CODE REVIEW

**Review by raphlinus — APPROVED** [review]
LGTM. I thought about doing this but was thinking more about code complexity than static size. The latter is important.

File: src/etc/unicode.py [hunk 1]
```diff
@@ -433,15 +444,49 @@ def emit_bool_trie(f, name, t_data, is_pub=True):
...
+    for (lo, hi) in t_data:
+        for cp in range(lo, hi + 1):
+            if int(cp / 64) >= len(chunks):
+                print(cp, int(cp / 64), len(chunks), lo, hi)
+            chunks[int(cp / 64)] |= 1 << (cp & 63)
+
+    pub_string = ""
+    if is_pub:
+        pub_string = "pub "
+    f.write("    %sconst %s: &'static super::SmallBoolTrie = &super::SmallBoolTrie {\n" % (pub_string, name))
```
Comment (frewsxcv) [inline src/etc/unicode.py]:
```
/checkout/src/etc/unicode.py:461: line longer than 100 chars
```

Comment (SimonSapin) [inline src/etc/unicode.py]: Fixed.

**Comment by nikomatsakis** [comment]
This seems good to me, but I feel like someone from @rust-lang/libs should weigh in.

r? @alexcrichton

**Comment by alexcrichton** [comment]
@bors: r+

Awesome wins, thanks @SimonSapin! Also thanks for the sanity check as well @raphlinus :)

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/issues/38851

**Comment by alexcrichton** [comment]
@bors: retry

* network failure

File: src/etc/unicode.py [hunk 2]
```diff
@@ -23,10 +23,7 @@
 # Since this should not require frequent updates, we just store this
 # out-of-line and check the unicode.rs file into git.
 
-import fileinput, re, os, sys, operator
-
-bytes_old = 0
-bytes_new = 0
+import fileinput, re, os, sys, operator, math
 
 preamble = '''// Copyright 2012-2016 The Rust Project Developers. See the COPYRIGHT
```
Comment (matthieu-m) [inline src/etc/unicode.py]: Ah, looks like the copyrights are lagging behind the new year.

Comment (SimonSapin) [inline src/etc/unicode.py]: Does it matter? Also, this PR is merged, so if you want something to happen you should open a new issue or PR.

**Pull request merged**